File: DSU.py
class dsu:
    def __init__(self, sz):
        self.sz = sz
        self.par = [i for i in range(sz)]
        self.rk = [0] * sz
    def find(self, a):
        if(self.par[a] == a):
            return a
        v = self.find(self.par[a])
        self.par[a] = v
        return v
    def merge(self, a, b):
        a = self.find(a)
        b = self.find(b)
        if(a == b):
            return False
        if(self.rk[a] > self.rk[b]):
            a, b = b, a
        self.par[a] = b
        self.rk[b] += 1
        return True

File: dsu.cpp
struct disjoint_set_union {
    vector<int> par, sz;

    disjoint_set_union() {}

    void resize(int size) {
        par.resize(size);
        sz.resize(size);
    }

    void make_set(int x) {
        par[x] = -1;
        sz[x] = 1;
    }

    int find(int x) {
        if (par[x] == -1) {
            return x;
        }
        par[x] = find(par[x]);
        return par[x];
    }

    bool merge(int a, int b) {
        int v, u;

        v = find(a), u = find(b);

        if (v == u) {
            return false;
        }

        if (sz[v] < sz[u]) {
            swap(v, u);
        }

        sz[v] += sz[u];
        par[u] = v;

        return true;
    }
};
